File: app/src/main/java/com/example/data/embedding/EmbeddingService.kt
package com.example.data.embedding

import android.util.Log
import com.example.data.XaiKeyStore
import com.example.data.remote.XAiApi
import com.example.data.remote.XAiEmbeddingRequest
import com.example.domain.model.Bookmark
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * Cloud embedding backend (xAI).
 *
 * xAI does not publish a fixed embedding-model name — availability is provisioned per key/team — so
 * this discovers the account's embedding model at runtime via [XAiApi.listEmbeddingModels] and only
 * falls back to [com.example.data.remote.GrokModels.EMBEDDING] if discovery yields nothing. The exact
 * server error (HTTP status + body) is captured in [lastError] and logged, so a "0 embedded" outcome
 * is explainable ("no embedding model on your xAI account" vs a real network/auth failure) instead of
 * a silent null.
 */
class EmbeddingService(private val xAiApi: XAiApi) : EmbeddingProvider {

    /** Human-readable reason the last embed attempt failed, for the UI to surface. Null on success. */
    @Volatile override var lastError: String? = null
        private set

    // Cache the discovered model id for the process lifetime so we don't re-list on every chunk.
    @Volatile private var cachedModelId: String? = null

    override fun isOnDevice(): Boolean = false

    override suspend fun embedDocument(bookmark: Bookmark): FloatArray? {
        // Chunk + mean-pool; cloud accepts much longer inputs than the seq256 on-device model.
        val vectors = EmbeddingText.chunksForDocument(bookmark, EmbeddingText.CLOUD_CHUNK_CHARS)
            .mapNotNull { embed(it) }
        return EmbeddingText.meanPool(vectors)
    }

    override suspend fun embedQuery(query: String): FloatArray? = embed(query)

    /**
     * Resolves the embedding model id for this account: cached → account-provisioned (discovered) →
     * the compile-time default. Discovery failures are non-fatal (we still try the default) but are
     * recorded so the caller can explain a subsequent failure.
     */
    private suspend fun resolveModelId(apiKey: String): String {
        cachedModelId?.let { return it }
        val discovered = try {
            xAiApi.listEmbeddingModels("Bearer $apiKey").firstModelId()
        } catch (e: retrofit2.HttpException) {
            Log.w("EmbeddingService", "embedding-models list HTTP ${e.code()}: ${errorBody(e)}")
            null
        } catch (e: Exception) {
            Log.w("EmbeddingService", "embedding-models list failed: ${e.message}")
            null
        }
        val id = discovered ?: com.example.data.remote.GrokModels.EMBEDDING
        cachedModelId = id
        if (discovered != null) Log.i("EmbeddingService", "Using discovered xAI embedding model: $id")
        return id
    }

    private suspend fun embed(text: String): FloatArray? = withContext(Dispatchers.IO) {
        try {
            if (!XaiKeyStore.isConfigured()) {
                lastError = "No xAI API key set (Settings → xAI API Key)."
                return@withContext null
            }
            val apiKey = XaiKeyStore.resolve()

            val request = XAiEmbeddingRequest(
                model = resolveModelId(apiKey),
                input = text.take(8000)
            )
            val response = xAiApi.createEmbeddings("Bearer $apiKey", request)
            val vector = response.data?.firstOrNull()?.embedding?.toFloatArray()
            if (vector == null) {
                lastError = "xAI returned no embedding — your account may not have an embedding model provisioned."
                Log.w("EmbeddingService", lastError!!)
            } else {
                lastError = null
            }
            vector
        } catch (e: retrofit2.HttpException) {
            val body = errorBody(e)
            lastError = when (e.code()) {
                404 -> "xAI has no embeddings endpoint/model for this key (404). Use On-device embeddings."
                401, 403 -> "xAI rejected the key for embeddings (HTTP ${e.code()})."
                else -> "xAI embeddings failed: HTTP ${e.code()} $body"
            }
            Log.e("EmbeddingService", "Embed HTTP ${e.code()}: $body")
            null
        } catch (e: Exception) {
            if (e is kotlinx.coroutines.CancellationException) throw e
            lastError = "xAI embeddings failed: ${e.message}"
            Log.e("EmbeddingService", "Embed failed: ${e.message}", e)
            null
        }
    }

    private fun errorBody(e: retrofit2.HttpException): String =
        runCatching { e.response()?.errorBody()?.string()?.take(300) }.getOrNull().orEmpty()
}

File: ios/Curio/Curio/Intelligence/EmbeddingService.swift
import Foundation
import os

/// Cloud embedding backend (xAI). Used as the fallback when on-device EmbeddingGemma is absent.
/// Ports `class EmbeddingService` from `data/embedding/EmbeddingService.kt`.
///
/// Caveat (from the Android source): as of 2026-06 xAI does NOT expose a public `/v1/embeddings` REST
/// endpoint (embeddings live only behind the Collections feature), so this call is best-effort and
/// usually returns `nil`. `EmbeddingProviderSelector` handles that by preferring on-device
/// EmbeddingGemma and the recency fallback in the chat path, so RAG keeps working regardless
/// (CONVENTIONS Networking cross-cutting "embedding endpoint best-effort (nil-tolerant)").
///
/// Resilient: logs + returns `nil` on any error, never throws (CONVENTIONS §3). Reuses the shared
/// `EmbeddingText` assembly so the cloud path embeds the *same* text as the on-device path.
final class EmbeddingService: EmbeddingProvider {

    private let xAiApi: XAiApi
    private let keyStore: XaiKeyStore
    private static let logger = Logger(subsystem: "com.curio.app", category: "EmbeddingService")

    /// Human-readable reason the last embed attempt failed, for the UI to surface. Nil on success.
    /// `nonisolated(unsafe)` (as with `entityStripRegex` above): this cloud path is best-effort and
    /// mirrors the Kotlin plain-class field; not worth an actor hop for a usually-nil fallback.
    private(set) nonisolated(unsafe) var lastError: String?

    /// Cache the discovered model id for the process lifetime.
    private nonisolated(unsafe) var cachedModelId: String?

    init(xAiApi: XAiApi, keyStore: XaiKeyStore = XaiKeyStore()) {
        self.xAiApi = xAiApi
        self.keyStore = keyStore
    }

    func isOnDevice() -> Bool { false }

    /// Chunk + mean-pool so long documents retain body content (mirrors the on-device path).
    /// Port of `embedDocument(bookmark)`.
    func embedDocument(_ bookmark: Bookmark) async -> [Float]? {
        var vectors: [[Float]] = []
        for chunk in EmbeddingText.chunksForDocument(bookmark, chunkChars: EmbeddingText.cloudChunkChars) {
            if let v = await embed(chunk) { vectors.append(v) }
        }
        return EmbeddingText.meanPool(vectors)
    }

    func embedQuery(_ query: String) async -> [Float]? {
        await embed(query)
    }

    /// Single best-effort embedding call. Port of the private `suspend fun embed(text)`:
    /// resolves the key, gates on `isConfigured()`, POSTs `input.take(8000)`, returns the first
    /// vector. Any error → `nil` and sets `lastError`.
    private func embed(_ text: String) async -> [Float]? {
        do {
            let apiKey = keyStore.resolve()
            guard keyStore.isConfigured() else {
                lastError = "No xAI API key set (Settings → xAI API Key)."
                return nil
            }

            let request = XAiEmbeddingRequest(
                model: try await resolveModelId(apiKey: apiKey),
                input: String(text.prefix(8000))
            )
            let response = try await xAiApi.createEmbeddings(
                authorization: "Bearer \(apiKey)",
                request: request
            )
            let vector = response.data?.first?.embedding
            if vector == nil {
                lastError = "xAI returned no embedding — your account may not have an embedding model provisioned."
                Self.logger.warning("\(self.lastError!, privacy: .public)")
            } else {
                lastError = nil
            }
            return vector
        } catch is CancellationError {
            return nil
        } catch let error as APIError {
            if case let .http(status, _, body) = error {
                let bodyText = String(data: body, encoding: .utf8).map { String($0.prefix(300)) } ?? ""
                lastError = switch status {
                case 404: "xAI has no embeddings endpoint/model for this key (404). Use On-device embeddings."
                case 401, 403: "xAI rejected the key for embeddings (HTTP \(status))."
                default: "xAI embeddings failed: HTTP \(status) \(bodyText)"
                }
            } else {
                lastError = "xAI embeddings failed: \(error)"
            }
            Self.logger.error("\(self.lastError!, privacy: .public)")
            return nil
        } catch {
            lastError = "xAI embeddings failed: \(error.localizedDescription)"
            Self.logger.error("\(self.lastError!, privacy: .public)")
            return nil
        }
    }

    /// Cached → account-provisioned (discovered) → compile-time default.
    private func resolveModelId(apiKey: String) async throws -> String {
        if let cached = cachedModelId { return cached }
        let discovered: String?
        do {
            discovered = try await xAiApi.listEmbeddingModels(authorization: "Bearer \(apiKey)").firstModelId
        } catch {
            Self.logger.warning("embedding-models list failed: \(error.localizedDescription, privacy: .public)")
            discovered = nil
        }
        let id = discovered ?? GrokModels.embedding
        cachedModelId = id
        if discovered != nil {
            Self.logger.info("Using discovered xAI embedding model: \(id, privacy: .public)")
        }
        return id
    }
}
